near-sdk/src/store/lookup_map/mod.rs

```diff
@@ -0,0 +1,533 @@
+    }
+}
+
+impl<K, V, H> LookupMap<K, V, H>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    /// Initialize a [`LookupMap`] with a custom hash function.
+    ///
+    /// # Example
+    /// ```
+    /// use near_sdk::hash::Keccak256;
+    /// use near_sdk::store::LookupMap;
+    ///
+    /// let map = LookupMap::<String, String, Keccak256>::with_hasher(b"m");
+    /// ```
+    pub fn with_hasher<S>(prefix: S) -> Self
+    where
+        S: IntoStorageKey,
+    {
+        Self {
+            prefix: prefix.into_storage_key().into_boxed_slice(),
+            cache: Default::default(),
+            hasher: Default::default(),
+        }
+    }
+
+    /// Overwrites the current value for the given key.
+    ///
+    /// This function will not load the existing value from storage and return the value in storage.
+    /// Use [`LookupMap::insert`] if you need the previous value.
+    ///
+    /// Calling `set` with a `None` value will delete the entry from storage.
+    pub fn set(&mut self, key: K, value: Option<V>) {
+        let entry = self.cache.get_mut(key);
+        match entry.get_mut() {
+            Some(entry) => *entry.value_mut() = value,
+            None => {
+                let _ = entry.set(CacheEntry::new_modified(value));
+            }
+        }
+    }
+
+    fn lookup_key<Q: ?Sized>(prefix: &[u8], key: &Q) -> LookupKey
+    where
+        Q: BorshSerialize,
+        K: Borrow<Q>,
+    {
+        // Concat the prefix with serialized key and hash the bytes for the lookup key.
+        let mut buffer = prefix.to_vec();
+        key.serialize(&mut buffer).unwrap_or_else(|_| env::panic_str(ERR_ELEMENT_SERIALIZATION));
+
+        H::hash(&buffer)
+    }
+}
+
+impl<K, V, H> LookupMap<K, V, H>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    fn deserialize_element(bytes: &[u8]) -> V {
+        V::try_from_slice(bytes).unwrap_or_else(|_| env::panic_str(ERR_ELEMENT_DESERIALIZATION))
+    }
+
+    fn load_element<Q: ?Sized>(prefix: &[u8], key: &Q) -> Option<V>
+    where
+        Q: BorshSerialize,
+        K: Borrow<Q>,
+    {
+        let storage_bytes = env::storage_read(&Self::lookup_key(prefix, key));
+        storage_bytes.as_deref().map(Self::deserialize_element)
+    }
+
+    /// Returns a reference to the value corresponding to the key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`BorshSerialize`] and [`ToOwned<Owned = K>`](ToOwned) on the borrowed form *must* match those for
+    /// the key type.
+    pub fn get<Q: ?Sized>(&self, k: &Q) -> Option<&V>
+    where
+        K: Borrow<Q>,
+        Q: BorshSerialize + ToOwned<Owned = K>,
+    {
+        //* ToOwned bound, which forces a clone, is required to be able to keep the key in the cache
+        let entry = self
+            .cache
+            .get(k.to_owned())
```

Comment:
thought: it's a bit sad that we `to_owned` on every access (instead of only on the *first* access), but this needs (unstable) raw_entry API to get fixed in a nice way.